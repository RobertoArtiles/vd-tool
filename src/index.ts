import path from 'path'
import execa from 'execa'

/**
 * spawn vd-tool process
 * 
 * @param args vd-tool cli args
 * @param options execa options
 */
export async function vdTool(
  args?: readonly string[],
  options?: execa.Options
) {
  const file = path.join(__dirname, '..', 'bin', 'vd-tool')
  if (!options) options = { stderr: process.stderr, stdout: process.stdout }
  Object.assign(options, { shell: true })
  return await execa(file, args, options)
}

type VdConvertOptions = {
  outDir?: string     // output dir path
  width?: number      // forces width in DP
  height?: number     // forces height in DP
  addHeader?: boolean // add AOSP header
  stdout?: execa.StdioOption // default to pipe
  stderr?: execa.StdioOption // default to pipe
}

/**
 * SVG to Vector Drawable conversion 
 * 
 * @param input file or directory path
 * @param options VdConvertOptions
 */
export async function vdConvert(
  input: string,
  options: VdConvertOptions = {}
) {
  const args = ['-c']
  const { outDir, width, height, addHeader, stderr, stdout} = options
  args.push('-in', String(input))
  if (outDir) args.push('-out', String(outDir))
  if (width) args.push('-widthDp', String(width))
  if (height) args.push('-heightDp', String(height))
  if (addHeader) args.push('--addHeader')
  return vdTool(args, {stderr, stdout})
}
