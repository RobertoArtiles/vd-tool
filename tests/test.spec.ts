import { vdConvert, vdTool } from '../src'
import path from 'path'
import fs from 'fs'

describe('tests', () => {
  it('converts svg', async () => {
    const fixture = path.join(__dirname, 'fixture', 'icon.svg')
    const outDir = path.join(__dirname, 'output')
    const result = await vdConvert(fixture, { outDir })
    expect(result.input).toMatch(/icon.svg/)
    expect(result.output).toMatch(/icon.xml/)
    expect(result.errors).toBeUndefined()
    expect(result.warnings).toBeUndefined()
    const xml = fs.readFileSync(
      path.join(__dirname, 'output', 'icon.xml'),
      'utf8'
    )
    expect(xml).toMatchSnapshot()
  }, 19999)

  it('reports errors and warnings', async () => {
    const fixture = path.join(__dirname, 'fixture', 'errors.svg')
    const outDir = path.join(__dirname, 'output')
    const result = await vdConvert(fixture, { outDir })
    expect(result.input).toMatch(/errors.svg/)
    expect(result.output).toMatch(/errors.xml/)
    expect(result.errors).toMatchInlineSnapshot(`
      Array [
        "ERROR @ line 9: Referenced id not found",
        "ERROR @ line 11: Referenced id not found",
      ]
    `)
    expect(result.warnings).toMatchInlineSnapshot(`undefined`)
    const xml = fs.readFileSync(
      path.join(__dirname, 'output', 'errors.xml'),
      'utf8'
    )
    expect(xml).toMatchSnapshot()
  }, 19999)
})
