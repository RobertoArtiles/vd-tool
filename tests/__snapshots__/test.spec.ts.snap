// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests converts svg 1`] = `
"<vector xmlns:android=\\"http://schemas.android.com/apk/res/android\\"
    android:width=\\"504dp\\"
    android:height=\\"672dp\\"
    android:viewportWidth=\\"5040\\"
    android:viewportHeight=\\"6720\\">
  <path
      android:pathData=\\"M44,6697c-2,-7 -3,-1003 -1,-2212 3,-2394 -1,-2230 57,-2500 52,-239 180,-558 309,-765 399,-642 1055,-1065 1806,-1165 158,-21 461,-21 620,0 561,74 1068,325 1462,723 369,374 599,832 690,1377 26,155 26,571 0,733 -24,147 -76,353 -122,477 -51,142 -182,403 -261,520 -354,530 -886,905 -1491,1051 -209,50 -347,65 -675,71l-308,6 -2,846 -3,846 -1038,3c-839,2 -1039,0 -1043,-11zM2610,2899c93,-25 144,-55 199,-114 213,-231 86,-593 -226,-647 -159,-27 -326,59 -401,208 -43,84 -52,150 -52,368l0,206 203,0c174,0 213,-3 277,-21z\\"
      android:fillColor=\\"#75be5e\\"
      android:strokeColor=\\"#00000000\\"/>
  <path
      android:pathData=\\"M2235,2807c-4,-8 -5,-95 -3,-193 3,-167 5,-182 28,-227 29,-58 86,-112 145,-138 55,-24 186,-28 235,-6 199,87 250,339 99,487 -77,74 -108,83 -316,88 -156,4 -184,2 -188,-11z\\"
      android:fillColor=\\"#75be5e\\"
      android:strokeColor=\\"#00000000\\"/>
  <path
      android:pathData=\\"M573,4493c3,-2209 3,-2218 24,-2320 117,-563 420,-1013 883,-1307 779,-495 1810,-372 2450,291 516,534 689,1302 450,1998 -152,442 -466,823 -875,1063 -174,103 -425,194 -640,233 -89,17 -168,22 -417,26l-308,5 0,-521 0,-521 193,0c237,0 307,-6 424,-35 309,-78 553,-315 650,-632 24,-79 26,-105 27,-248 1,-138 -2,-171 -22,-238 -91,-315 -310,-541 -622,-642 -70,-23 -109,-28 -221,-33 -218,-7 -374,34 -545,147 -190,125 -325,316 -386,543l-23,83 -2,2163 -3,2162 -520,0 -520,0 3,-2217z\\"
      android:fillColor=\\"#ffffff\\"
      android:strokeColor=\\"#00000000\\"/>
  <path
      android:pathData=\\"M2235,2807c-4,-8 -5,-95 -3,-193 3,-167 5,-182 28,-227 29,-58 86,-112 145,-138 55,-24 186,-28 235,-6 199,87 250,339 99,487 -77,74 -108,83 -316,88 -156,4 -184,2 -188,-11z\\"
      android:fillColor=\\"#ffffff\\"
      android:strokeColor=\\"#00000000\\"/>
</vector>
"
`;

exports[`tests reports errors and warnings 3`] = `
"<vector xmlns:android=\\"http://schemas.android.com/apk/res/android\\"
    android:width=\\"80dp\\"
    android:height=\\"80dp\\"
    android:viewportWidth=\\"80\\"
    android:viewportHeight=\\"80\\">
  <path
      android:pathData=\\"M31.84,46.02L38.28,46.02L38.28,47.16L30,47.16L30,46.24L33.74,41.94C34.8333,40.6733 35.56,39.7 35.92,39.02C36.28,38.34 36.46,37.5933 36.46,36.78C36.46,35.9 36.2267,35.2533 35.76,34.84C35.2933,34.4267 34.5733,34.22 33.6,34.22C33.1067,34.22 32.6067,34.2633 32.1,34.35C31.5933,34.4367 31.0133,34.58 30.36,34.78L30.36,33.58C30.96,33.38 31.53,33.2333 32.07,33.14C32.61,33.0467 33.16,33 33.72,33C35.0933,33 36.13,33.31 36.83,33.93C37.53,34.55 37.88,35.46 37.88,36.66C37.88,37.58 37.6833,38.4267 37.29,39.2C36.8967,39.9733 36.1067,41.0533 34.92,42.44L31.84,46.02ZM50.46,43.14L50.46,44.28L48.36,44.28L48.36,47.16L46.98,47.16L46.98,44.28L40.48,44.28L40.48,43.22C41.2667,41.38 42.1533,39.6433 43.14,38.01C44.1267,36.3767 45.2533,34.7733 46.52,33.2L48.36,33.2L48.36,43.14L50.46,43.14ZM41.88,43.14L46.98,43.14L46.98,34.48C45.94,35.84 45.0033,37.23 44.17,38.65C43.3367,40.07 42.5733,41.5667 41.88,43.14Z\\"
      android:fillColor=\\"#FFFFFF\\"/>
</vector>
"
`;
